### ISSUE TITLE
[24.0rc4] Error: Cannot resolve -bind address: 'bitcoind:8334=onion'

### ISSUE THREAD

**Issue by schildbach** [bitcoin/bitcoin#26484]
In my quest to finally fix #25094, I figured I need to bind the onion listen port explicitly to the external interface (`bitcoind`), not to "all interfaces" (`[::]`). So I'm using the argument

```
-bind=bitcoind:${BITCOIND_TOR_PORT}=onion
```

`/etc/hostname` is set to `bitcoind` and the `hostname` command yields the expected result. Also, both `nslookup` and `ping` can resolve the hostname to the external IP.

Note: I cannot use the numerical IP, because Docker assigns IP addresses dynamically on each invocation.

**Expected behavior**

I'd expect bitcoind to also resolve the hostname to my external IP, just like ping and nslookup do.

**Actual behavior**

In short, it dies.

```
bitcoind_1          | Error: Cannot resolve -bind address: 'bitcoind:8334=onion'
bitcoind_1          | Shutdown: In progress...
bitcoind_1          | Error: Cannot resolve -bind address: 'bitcoind:8334=onion'
bitcoind_1          | Shutdown: done
```

**System information**

bitcoind 24.0rc4 (self built), Docker+Compose from Ubuntu 22.04.1.

**Comment by fanquake** [comment]
@vasild you might have some insight here?

**Comment by schildbach** [comment]
Fwiw, if I exec into my container, replace the `bitcoind` string manually with the external IP address and launch bitcoind again, it works – until I restart my container, which assigns a new IP. So it's really just the resolving of the address that doesn't work.

**Comment by schildbach** [comment]
I see these options for configuring the bind port of a Tor hidden service:

1. Use wildcard addresses (`0.0.0.0` and/or `[::]`). This binds correctly, but it seems bitcoind is picking one of them and using that (via Tor control port) for configuration of the target address of the hidden service. This fails, because the wildcard address has a different meaning to the Tor container. To fix this issue, bitcoind could offer a separate configuration option for the target address (which would support hostnames, see below). This issue is tracked in #25094.
2. Use hostnames to bind to a specific external interface, which can be reached via that name or resolved IP from the Tor container. As stated above, bitcoind currently does not support this. I don't get the security argument – it's up to the user to trust their name resolution and configure bitcoind either by hostname or IP as desired.
3. Use fixed numerical IP, which can be reached from the Tor container. However, IPs are typically assigned dynamically (e.g. by Docker). This makes configuration overly complicated, as your workaround clearly shows. For example, Docker containers use arbitrary values in `/etc/hosts` and `/etc/hostname` by default, so your workaround wouldn't work out of the box. I'd rather use `nslookup`, if I were to go that route.

As you can see, currently none of the options work with the configuration of a Tor hidden service, managed by bitcoind via control port. Currently I can only revert back to the old method of managing the hidden service in the Tor container.

**Comment by vasild** [comment]
@schildbach is this behavior something new? Is it a regression since an older version of `bitcoind` which worked "better"?

> ... I don't get the security argument ...

If you feel that this should be changed, then you should open an issue for that or even a pull request that flips `false` to `true` on this line:

https://github.com/bitcoin/bitcoin/blob/7ef730ca84bd71a06f986ae7070e7b2ac8e47582/src/init.cpp#L1720

Don't be deterred by my opinion - it is just one. Maybe others would think it is ok to change it.

FWIW I checked the behavior of:
* Tor (`SOCKSPort google.com:9050`)
* Apache (`Listen google.com:80`) and
* sshd (`ListenAddress google.com)`

All of them do resolve the name to an IP address and try to bind to it (and of course failed miserably with my examples).

I guess that `bitcoind` only supports option 3 currently. Somehow is has been assumed that the operator knows their IP address. I understand that Docker makes this difficult. I am not very familiar with it, but does it not provide the option to run some custom setup scripts inside the host OS after setting it up and the IP address to be in some environment variable, available to such a script?

> Currently I can only revert back to the old method of managing the hidden service in the Tor container.

What is that old method?

**Comment by schildbach** [comment]
> What is that old method?

Rather than letting bitcoind manage the Tor hidden servce via the control port, you'd configure it with tor (look up `HiddenServiceDir`, `HiddenServicePort`). In this case, you configure the target address within tor, and tor has no regrets using hostname resolution. The main downside is the hidden service private key is then managed by Tor.

Both approaches are described in `doc/tor.md`, however it is missing documentation of the various pitfalls which come with `-torcontrol` mode.

> Is this behavior something new?

Well, I'm trying to get it to work for months now. 24.0 had some other bug fixed that prevented `-torcontrol` mode, but now I'm running into more problems.

**Comment by 1440000bytes** [comment]
@michaelfolkson can you fix this?

**Comment by fanquake** [comment]
Not completely sure what the status is here, but I've moved this to the 24.1 milestone.

**Comment by vasild** [comment]
To summarize, @schildbach, correct me if I am wrong - there are two issues:

1. `bitcoind` does not resolve the argument to `-bind=...`, so an IP address must be provided by the user. It is a feature request to change the behavior.
2. If `bitcoind` listens on `0.0.0.0` or `::` for incoming tor connections, then it would tell the tor daemon to connect to that address. This is a bug that could be resolved by https://github.com/bitcoin/bitcoin/pull/22729#issuecomment-927802257.

It has been like this since the start and is not a regression from a previous version. There is no patch for either one at the moment.

**Comment by schildbach** [comment]
> f you feel that this should be changed, then you should open an issue for that or even a pull request that flips `false` to `true` on this line:
> 
> https://github.com/bitcoin/bitcoin/blob/7ef730ca84bd71a06f986ae7070e7b2ac8e47582/src/init.cpp#L1720

Actually flipping that bool doesn't work. It seems to try to resolve the entire line, e.g.
```
Error: Cannot resolve -bind address: 'bitcoind:38334=onion'
```

**Comment by schildbach** [comment]
I think the cleanest way to fix this is keep the `-bind` logic as is, but introduce a new optional argument `-onionporttarget` that specifies the string that is sent to the tor control port. The default would be taken from the bind as it is done today. That way, tor can resolve the IP and we can specifiy similar to:

```
-bind=0.0.0.0:8334=onion
-onionporttarget=bitcoind:8334
```

**Comment by schildbach** [comment]
So should I try switching the bool in line 1853 rather than line 1841?

**Comment by vasild** [comment]
Yes, line `1841` is for when there is no `=` inside the argument, like in `-bind=1.2.3.4:8333`. Line `1853` is for when there is `=`, like `-bind=1.2.3.4:8333=onion`.

**Comment by dayer3** [comment]
Hi all,
Sorry, so what is the recommended workaround for creating a Bitcoin Core onion service if Tor service is running on a different host or container, without following the _Manually create a Bitcoin Core onion service_ way and bind=a.b.c.d:8334=onion?
Best regards

**Comment by vasild** [comment]
@dayer3, so Tor daemon is running on another machine and you want to specifically use `-bind=hostname:8334=onion` in `bitcoind` config and not `-bind=a.b.c.d:8334=onion`? I am not sure if that is possible. This comment contains some workaround:

https://github.com/bitcoin/bitcoin/issues/26484#issuecomment-1311427984
